lib/puppet/property/keyvalue.rb

```diff
@@ -53,11 +72,33 @@ def process_current_hash(current)
       def should
         return nil unless @should
 
-        members = hashify(@should)
+        members = hashify_should
         current = process_current_hash(retrieve)
 
         #shared keys will get overwritten by members
-        current.merge(members)
+        should_value = current.merge(members)
+
+        # Figure out the keys that will actually change in our Puppet run.
+        # This lets us reduce the verbosity of Puppet's logging for instances
+        # of this class when we want to.
+        #
+        # NOTE: This feature only makes sense when current + should_value will
+        # always have the same keys (i.e. when the keys of our KeyValue property
+        # will never change). If this precondition isn't met, then be sure that
+        # self.class.log_only_changed_keys is set to false so we don't actually
+        # do anything with this value.
+        #
+        # NOTE: We use ||= here because we only need to compute the changed_keys
+        # once (since this property will only be synced once).
+        #
+        @changed_or_new_keys ||= should_value.keys.select do |key|
+          # A key that's exclusively in should was added, so it is
+          # automatically changed.
+          next true unless current.key?(key)
```

Comment:
I'm not familiar with this `next true...` syntax. What does it do?

Reply:
Oh, I think I get it. It's like an early return for a block? That's gross

Reply:
Yeah, next is not a very good name.

Reply:
`!current.key?(key) || current[key] != should_value[key]` maybe?

Reply:
Yes, that is better. I realized my NOTE is contradicting itself (the first one) since the first part is saying there's a difference between current and should. Will reword that.